### ISSUE TITLE
Gvim visual artifacts on MS Windows

### ISSUE THREAD

**Issue by harg** [vim/vim#8532]
**Describe the bug**
Visual artifacts/glitchs appear when inputing some characters on MS Windows Gvim

**To Reproduce**
1. Run gvim `gvim --clean`
2. Change font to consolas `:set guifont=Consolas:h10.5`
2. Change colorscheme to slate `:colo slate`
3. Type '@@@@@@' 
4. go to normal mode and open a new vertical split `:vsp`
5. go to insert mode and press return several times 
4. You can see the visual glitchs in red

**Screenshot**
![image](https://user-images.githubusercontent.com/34398/124983180-f13c0e00-e037-11eb-8a39-7f5a51307a29.png)

**Expected behavior**
No visual artifacts/glitchs when inputing characters

**Environment**
 - Vim version : 8.2.3119 x64
 - OS: Windows 10
 - gvim

**Additional context**
It can be reproduced on several windows machines.
Artifacts are gone with a `:redraw`  or Ctrl-L
`:set renderoptions=type:directx` does not fix the problem.
Same problem with Vim x86.
Same problem with different widely used fonts (Fira Code Mono, Monaco, ...).

related bug : https://github.com/vim/vim/issues/5549

**Comment by nanis** [comment]
I tried this and was not able to reproduce. `gvim` was built using MS Visual Studio 19 tools with the command:

```
make -f Make_mvc.mak CPU=AMD64 CPUNR=sse2 OPTIMIZE=SPEED FEATURES=HUGE NETBEANS=no GETTEXT=no CSCOPE=no XPM=no DIRECTX=yes GUI=yes COLOR_EMOJI=yes MBYTE=yes
```

and

```
:version
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jul 15 2021 21:07:22)
MS-Windows 64-bit GUI version
Included patches: 1-3171  
```

**Comment by chrisbra** [comment]
Could this be a font issue?

**Comment by nanis** [comment]
@chrisbra I did try with Consolas as well as the other fonts mentioned. I am on Windows 10 Pro 21H1 with ClearType enabled which I forgot to mention before.

**Comment by harg** [comment]
I use the build from : https://github.com/vim/vim-win32-installer/releases/

**Comment by nanis** [comment]
Thank you @harg. I was not aware of that repo.

I downloaded the x64 ZIP file from here https://github.com/vim/vim-win32-installer/releases/tag/v8.2.3182 and ran `gvim --clean` in the extracted directory.

![Screenshot](https://i.imgur.com/fBIqRqj.png)

**Comment by chrisbra** [comment]
So, this can be closed?

**Comment by nanis** [comment]
The artifacts seem to be caused by something other than `gvim` code. It is possible there might be something the code can do to prevent these artifacts, but without further information on the specific set of circumstances that trigger the artifacts, I think the ticket can be closed.

**Comment by gdupras** [comment]
@nanis, the visual glitches *are* visible in the image you posted. Look carefully. There are two red spots in the left split:
- On the right side of the last `@` of the first line.
- On the right side of the `@` of the last line.

Also, you didn't follow the steps in the OP exactly. To be fair, they could have been a little clearer. In step 6, with the cursor on the last `@`, press `i` to enter insert mode. The cursor should now be between the fifth and the sixth `@`. Pressing `Enter` will then move down that last `@`.

By following these steps, I end up with the exact same image @harg posted. I'm using the latest nightly build (8.2.3182 x64) on Windows 10.

**Comment by nanis** [comment]
@gdupras Yes, you are correct. With your instructions, I can also reproduce. Also, congrats on spotting the red pixels in the original image I posted. I could not tell those apart from the usual ClearType fuzz without you specifically pointing them out.

@harg @chrisbra Apologies for the inconvenience. You can consider this replicated.

![screenshot](https://i.imgur.com/lgeJvQD.png)

**Comment by nanis** [comment]
Without ClearType, the artifacts are more prominent, but, obviously, the coloration is gone.

![screenshot with no cleartype](https://i.imgur.com/NoBC5QR.png)

**Comment by harg** [comment]
Sorry if my instructions weren't clear enough...  :(

**Comment by nanis** [comment]
> Sorry if my instructions weren't clear enough... :(

Let's just say I understood them correctly after the second explanation :-)

**Comment by char101** [comment]
The artifacts comes from antialiasing which are not cleared when clearing the character since with antialiasing, the character drawn width > the character width.

As can be seen in the image below, the artifact matches the rgb antialiasing.

![2021-08-06 09_55_53- No Name  + - GVIM11](https://user-images.githubusercontent.com/71255/128449456-4b140325-8a6b-464a-a6fa-a49fd9ecc277.png)

**Comment by harg** [comment]
Hello, any news here?
Somewhat it was partially fixed here when you don't have splits and for artifacts drawn at right side of char : https://github.com/vim/vim/issues/5549 

it still occurs on both side of "wide" char when you have splits and on left side when you don't have split.

**Comment by char101** [comment]
If you compile vim yourself, there is a patch here

https://groups.google.com/g/vim_dev/c/_a13OUtm4fE?pli=1

![2022-06-25 08_43_11- No Name  + - GVIM137](https://user-images.githubusercontent.com/71255/175753580-ba6db874-aabc-4a2a-9571-a0746000a38e.png)

**Comment by brammool** [comment]
> If you compile vim yourself, there is a patch here
> 
> https://groups.google.com/g/vim_dev/c/_a13OUtm4fE?pli=1
> 
> ![2022-06-25 08_43_11- No Name  + - GVIM137](https://user-images.githubusercontent.com/71255/175753580-ba6db874-aabc-4a2a-9571-a0746000a38e.png)

That was for Windows XP, let us know if this is still a valid fix.

The problem is that the patch is incomplete.  It always does an extra
redraw.  There should be a condition to only do this when it is needed.

-- 
I learned the customs and mannerisms of engineers by observing them, much the
way Jane Goodall learned about the great apes, but without the hassle of
grooming.
				(Scott Adams - The Dilbert principle)

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by char101** [comment]
I have been using it in Windows 10 for a long time and so far I think it is working fine (I haven't notice any subpixel artifact).

This is the latest patch from my hg series

```patch
# HG changeset patch
# Parent  2ffbf8961d35eb8dc23888e124dffe5b0d2c03ec

diff -r 2ffbf8961d35 src/screen.c
--- a/src/screen.c	Sun Aug 21 17:29:39 2022 +0700
+++ b/src/screen.c	Sun Aug 21 17:29:40 2022 +0700
@@ -457,6 +457,10 @@
     int		    clear_next = FALSE;
     int		    char_cells;		// 1: normal char
 					// 2: occupies two display cells
+#ifdef FEAT_GUI
+    int		    changed_this;   /* bool: is character changed? */
+    int		    changed_next;   /* bool: changed_this for next character */
+#endif
 
     // Check for illegal row and col, just in case.
     if (row >= Rows)
@@ -513,6 +517,9 @@
 #endif
 
     redraw_next = char_needs_redraw(off_from, off_to, endcol - col);
+#ifdef FEAT_GUI
+    changed_next = redraw_next;
+#endif
 
     while (col < endcol)
     {
@@ -529,14 +536,14 @@
 	// If the next character was bold, then redraw the current character to
 	// remove any pixels that might have spilt over into us.  This only
 	// happens in the GUI.
+	changed_this = changed_next;
+	changed_next = redraw_next;
+	// Redraw the current character to remove any pixels that might have
+	// spilt over into us.  This only happens in the GUI, and only when
+	// subpixel rendering is used. We can only detect if the GUI is used, so
+	// we should always redraw.
 	if (redraw_next && gui.in_use)
-	{
-	    hl = ScreenAttrs[off_to + char_cells];
-	    if (hl > HL_ALL)
-		hl = syn_attr2attr(hl);
-	    if (hl & HL_BOLD)
-		redraw_this = TRUE;
-	}
+	    redraw_this = TRUE;
 #endif
 #ifdef FEAT_PROP_POPUP
 	if (blocked_by_popup(row, col + coloff))
@@ -668,6 +675,11 @@
 		    redraw_next = TRUE;
 	    }
 #endif
+
+#ifdef FEAT_GUI
+	    if (gui.in_use && changed_this)
+		redraw_next = TRUE;
+#endif
 	    ScreenAttrs[off_to] = ScreenAttrs[off_from];
 	    ScreenCols[off_to] = ScreenCols[off_from];
```

**Comment by brammool** [comment]
That looks like a reasonable approach.  Let me include that but with FEAT_GUI_MSWIN.

**Comment by mao-yining** [comment]
Sounds this is fixed. @chrisbra

**Comment by chrisbra** [comment]
Closing then
